Fix encryption:fix-encrypted-version command when encrypted is set to 0

come-nc opened the pull request:
Fix https://github.com/nextcloud/server/issues/31874

come-nc committed:
Improve typing in FixEncryptedVersion command

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix encryption:fix-encrypted-version command when encrypted is set to 0

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

PVince81 commented on apps/encryption/lib/Command/FixEncryptedVersion.php:
```diff
@@ -185,6 +175,22 @@ private function verifyFileContent($path, OutputInterface $output, $ignoreCorrec
+				}
+				$encryptedVersion = $fileInfo->getEncryptedVersion();
+				$stat = $this->view->stat($path);
+				if (($encryptedVersion == 0) && isset($stat['hasHeader']) && ($stat['hasHeader'] == true)) {
```
we should only do this if an explicit command line flag is passed

otherwise there's a risk a false positives in case a file would look like it has a header with the same bytes but isn't actually a Nextcloud encrypted file (probability is likely 0.0000001% though)

PVince81 replied:
the decryption will fail anyway if it was a false positive

PVince81 reviewed (approved):
:+1:

PVince81 commented:
/backport to stable24

PVince81 commented:
@come-nc please fix the test

come-nc committed:
Adapt test to fixed command output.

No user and non-existing user are now correctly treated as two separated
 cases

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
> @come-nc please fix the test

fixed

artonge commented on apps/encryption/lib/Command/FixEncryptedVersion.php:
```diff
@@ -212,18 +218,19 @@ private function verifyFileContent($path, OutputInterface $output, $ignoreCorrec
 		}
 		$fileId = $fileInfo->getId();
+		if ($fileId === null) {
+			$output->writeln("<warning>File info contains no id for file: \"$path\"</warning>");
```
Seems more appropriate:

```suggestion
			$output->writeln("<warning>Unknown file: \"$path\"</warning>");
```

come-nc replied:
Well, this is a case that should not happen I think, I added the test to please psalm, and in the case it happens I would prefer the warning to be quite precise to ease debugging. So yeah the message is technical, but so it the error case.

artonge commented on apps/encryption/lib/Command/FixEncryptedVersion.php:
```diff
@@ -185,6 +175,22 @@ private function verifyFileContent($path, OutputInterface $output, $ignoreCorrec
 			if (\fread($handle, 9001) !== false) {
+				$fileInfo = $this->view->getFileInfo($path);
+				if (!$fileInfo) {
+					$output->writeln("<warning>File info not found for file: \"$path\"</warning>");
```
Seems more appropriate:

```suggestion
					$output->writeln("<warning>Unknown file: \"$path\"</warning>");
```

enoch85 reviewed (approved):
Tested and works!
